### PR TITLE
Webpack 5 compat

### LINKED ISSUE #12: Webpack 5 shows decprication warning for compilation assets
When using _inline-source-webpack-plugin_ 1.4.1 with Webpack 5.28.0, I am getting this warning:

```
[DEP_WEBPACK_COMPILATION_ASSETS] DeprecationWarning: Compilation.assets will be frozen in future, all modifications are deprecated.
BREAKING CHANGE: No more changes should happen to Compilation.assets after sealing the Compilation.
        Do changes to assets earlier, e. g. in Compilation.hooks.processAssets.
        Make sure to select an appropriate stage from Compilation.PROCESS_ASSETS_STAGE_*.```

### PR DESCRIPTION
**Pull request by duncanbeevers**
Use the [processAssets compilation hook](https://webpack.js.org/api/compilation-hooks/#processassets) to modify the compilation assets list before the compilation is sealed.

Closes #12

I wrote this in a way such that it works with webpack 4 and webpack 5
I broke the PR into two commits to make it simple to test it against both versions.

### CODE REVIEW

File: test/index.spec.js [hunk 1]
```diff
@@ -79,16 +79,10 @@ function testInlineSourceWebpackPlugin(
...
-            return result.test(content);
-          } else {
-            return content.indexOf(result) > -1;
-          }
-        })
-      ).toBeTruthy();
+      const content = fs.readFileSync(path.join(__dirname, `../dist/${outputFile}`)).toString();
+      expectResults.forEach(result => {
+        expect(content).toMatch(result);
+      });
```
Comment (duncanbeevers) [inline test/index.spec.js:85]: The compilation source no longer contains the content after emit.

https://webpack.js.org/blog/2020-10-10-webpack-5-release/#sizeonlysource-after-emit

> webpack now replaces the Sources in Compilation.assets with SizeOnlySource variants to reduce memory usage.

File: test/index.spec.js [hunk 2]
```diff
@@ -110,11 +104,11 @@ describe('InlineSourceWebpackPlugin', () => {
         // inline.js
         `<script>function Person(){}Person.prototype.sayHello=function(){console.log("[inline]:","hello world!")},(new Person).sayHello();</script>`,
         // webpack runtime file
-        `window.webpackJsonp=window.webpackJsonp||[]`,
+        `webpack`,
```
Comment (duncanbeevers) [inline test/index.spec.js:107]: The webpack runtime chunk has changed between webpack versions. This very relaxed test catches both old and new runtimes, but may be over-broad.

File: test/index.spec.js [hunk 3]
```diff
@@ -110,11 +104,11 @@ describe('InlineSourceWebpackPlugin', () => {
...
         // bundle.js
         `console.log("This file is build by webpack.But InlineSourceWebpackPlugin will embed it into html file.")`,
         // appended by html-webpack-plugin
-        /<script src="\/inline-source-webpack-plugin\/dist\/index\.\w+\.js"><\/script>/
+        /<script (?:defer="defer" )?src="\/inline-source-webpack-plugin\/dist\/index\.\w+\.js"><\/script>/
```
Comment (duncanbeevers) [inline test/index.spec.js:111]: New versions of `html-webpack-plugin` include the `defer="defer"` attribute on the entry script tag.
This test adds an optional clause to the regexp to allow the entry script tag to include, or not include this attribute.

File: test/index.spec.js [hunk 4]
```diff
@@ -190,8 +184,6 @@ describe('InlineSourceWebpackPlugin', () => {
     testInlineSourceWebpackPlugin({
       webpackConfig,
       done,
-      expectOutputFiles: [/index\.html$/, /index\.\w+\.js$/, /runtime\.\w+\.js$/, /bundle\.\w+\.js$/],
-      expectResults: ['<!--inline-source-webpack-plugin-->'],
```
Comment (duncanbeevers) [inline test/index.spec.js:194]: When `noAssetMatch` is set to `'error'` webpack 5 doesn't create the `dist/` folder.
I didn't drill too far into why this wasn't built as the hard error indicates a failed build.

Hopefully this isn't a deal-breaker.

File: test/index.spec.js [hunk 5]
```diff
@@ -79,16 +79,10 @@ function testInlineSourceWebpackPlugin(
...
-          if (result instanceof RegExp) {
-            return result.test(content);
-          } else {
-            return content.indexOf(result) > -1;
-          }
-        })
-      ).toBeTruthy();
+      const content = fs.readFileSync(path.join(__dirname, `../dist/${outputFile}`)).toString();
+      expectResults.forEach(result => {
+        expect(content).toMatch(result);
```
Comment (duncanbeevers) [inline test/index.spec.js:84]: The [`toMatch`](https://jestjs.io/docs/expect#tomatchregexp--string) matcher accepts a string or regexp and does a substring match when given a string.

I can roll this back, but I found the resulting failing assertions much easier to read and debug than the simple boolean pass/fail.

File: index.js [hunk 6]
```diff
@@ -157,10 +164,19 @@ class InlineSourceWebpackPlugin {
...
         });
       }
-      compiler.hooks.emit.tapAsync(name, (compilation, callback) => {
-        this._deleteAsset(compilation);
-        callback && callback();
-      });
+      if (Compilation && Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE) {
+        compiler.hooks.thisCompilation.tap(name, compilation => {
+          compilation.hooks.processAssets.tap(
+            { name, stage: Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE },
```
Comment (duncanbeevers) [inline index.js:170]: There are a number of stages at which to hook into the pre-sealing asset processing phase; https://github.com/webpack/webpack/blob/d009b13aa231957936c394bb2cf7df275709193a/lib/Compilation.js#L4440-L4519

The one that seemed most appropriate was `PROCESS_ASSETS_STAGE_OPTIMIZE_INLINE`.
However, that phase runs *before* the `beforeEmit` hook fires. The `beforeEmit` / `_process` work is necessary to build up the list of assets-to-delete.

Instead, I used `PROCESS_ASSETS_STAGE_SUMMARIZE` which runs *after* the `beforeEmit` hook fires, and therefore has access to the final list of assets-to-delete.

**Review by KyLeoHC — APPROVED** [review]
Thanks for your PR.

**Pull request merged**